worker/master/helmapp/app.go

```diff
@@ -0,0 +1,317 @@
+// must be obtained first.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+package helmapp
+
+import (
+	"context"
+	"path"
+
+	"github.com/goodrain/rainbond/pkg/apis/rainbond/v1alpha1"
+	"github.com/goodrain/rainbond/pkg/generated/clientset/versioned"
+	"github.com/goodrain/rainbond/pkg/helm"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"helm.sh/helm/v3/pkg/release"
+	"helm.sh/helm/v3/pkg/storage/driver"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+)
+
+// App represents a helm app.
+type App struct {
+	ctx            context.Context
+	log            *logrus.Entry
+	rainbondClient versioned.Interface
+	recorder       record.EventRecorder
+
+	helmApp         *v1alpha1.HelmApp
+	originalHelmApp *v1alpha1.HelmApp
+
+	name         string
+	namespace    string
+	templateName string
+	version      string
+	repoName     string
+	repoURL      string
+	overrides    []string
+	revision     int
+	chartDir     string
+
+	helmCmd *helm.Helm
+	repo    *helm.Repo
+}
+
+// Chart returns the chart.
+func (a *App) Chart() string {
+	return a.repoName + "/" + a.templateName
+}
+
+// NewApp creates a new app.
+func NewApp(ctx context.Context, kubeClient clientset.Interface, rainbondClient versioned.Interface, helmApp *v1alpha1.HelmApp, repoFile, repoCache string) (*App, error) {
+	helmCmd, err := helm.NewHelm(helmApp.GetNamespace(), repoFile, repoCache)
+	if err != nil {
+		return nil, err
+	}
+	repo := helm.NewRepo(repoFile, repoCache)
+	log := logrus.WithField("HelmAppController", "Reconcile").WithField("Namespace", helmApp.GetNamespace()).WithField("Name", helmApp.GetName())
+
+	return &App{
+		ctx:             ctx,
+		log:             log,
+		recorder:        createRecorder(kubeClient, helmApp.Name, helmApp.Namespace),
+		rainbondClient:  rainbondClient,
+		helmApp:         helmApp.DeepCopy(),
+		originalHelmApp: helmApp,
+		name:            helmApp.GetName(),
+		namespace:       helmApp.GetNamespace(),
+		templateName:    helmApp.Spec.TemplateName,
+		repoName:        helmApp.Spec.AppStore.Name,
+		repoURL:         helmApp.Spec.AppStore.URL,
+		version:         helmApp.Spec.Version,
+		revision:        helmApp.Spec.Revision,
+		overrides:       helmApp.Spec.Overrides,
+		helmCmd:         helmCmd,
+		repo:            repo,
+		chartDir:        path.Join("/tmp/helm/chart", helmApp.Namespace, helmApp.Name, helmApp.Spec.Version),
```

Comment:
What Is chartDir?

Reply:
chartDir is the cache directory of helm charts